### DISCUSSION TITLE
sile vs context lmtx vs mckenna's jsbox

### DISCUSSION THREAD

**Discussion by goyalyashpal** [sile-typesetter/sile#1905] [General]
hi!

i had several doubts and questions... i held on to these for quite a long time, but couldn't reach a conclusion & as digital (& programmatic) typesetting is fascinating & very personal for me, i couldn't give up on it either. 

so, finally thought to ask. hope u won't mind.

not considering https://github.com/sile-typesetter/sile/discussions/1698 as that question has a particular goal in mind (abstraction: lyx/md), i want an overview, a map if you will of how-&-why things are.

----
## A.

1. is [libtexpdf] as a standalone library merged upstream now? as the potent mentioned by Simon in this [2015 FOSDEM talk][5BIP_N9qQm4]
2. the [context-LMTX] from 2019 now has several overlapping areas with sile:
    * Lua & XML support
    * grids & layers 
    * lack of latex packages support 😝
3. so, what is the difference in context-lmtx and sile now? the major one seems to be:
    * frame balancing, 
    * citations/footnotes (https://github.com/sile-typesetter/sile/discussions/1316#discussioncomment-2239399)
4. reading the [sile-0.14.13] manual, [examples], the converters (5.3.3), dropcaps (5.4.1), leaders (5.6.1) & snake justification (lol), gutenberg (5.4.6), flowed text, multiple scripts etc. seem interesting - i don't know if they're possible in context or not

is there any chance that context-lmtx & sile may join hands? would it be beneficial? why or why not?

[libtexpdf]: https://github.com/sile-typesetter/libtexpdf "libtexpdf dvipdfm extract"
[sile-0.14.13]: https://sile-typesetter.org/manual/sile-0.14.13.pdf "sile book v0.14.13"
[examples]: https://sile-typesetter.org/examples/ "Examples"

[5BIP_N9qQm4]: https://www.youtube.com/watch?v=5BIP_N9qQm4 "Introducing SILE | 2015.03"
[context-lmtx]: https://tug.org/TUGboat/tb40-1/tb124hagen-lmtx.pdf "ConTeXt LMTX | 2019 Hans Hagen"

----
## B.

<details><summary>Similar to how <i>(click to expand)</i></summary>

* kotlin compiles to java bytecode; hence can use existing java libs
* python extensions (pypy, mypy, hylang, etc) compile to python bytecode; hence can use python infra
* zig has same ABI as C; hence can use existing C infra
* typescript compiles to javascript; hence can use existing JS infra

</details>

... why not such similar thing can be done for TeX? 
* isn't there some common thing that these all things produce so that these new tex-derivation things can interoperate as well? 
* like, there has to have _some_ common thing which all these write while making a pdf (or other output forms), so, why can't these be like, used side-by-side? 
* say: sile: i have compiled the part upto here, hey latex package, now you spit your output. 

as i don't have much problem trying new things, new syntax, new paradigm, whatever - it's just that i _need_ packages support. and lack of interoperability is a _huge_ hurdle in this aspect.

it doesn't matter what anyone feels about legacy systems - the libraries developed there _should be_ available in newer systems, that's how new things gain traction; and _that_ is how we progress. the 4 above examples of new languages illustrate that.

developing new things from scratch alone won't help much in general area outside of niches.

----
## C.

[tectonic] is a builder for tex? what is this [jsbox][jsbox-overview] thing? 
can any of these or other such thing help sile in any way? or allow it to take benefit of existing latex packages?

[tectonic]: https://tectonic-typesetting.github.io/ "tectonic typesetting"
[jsbox-overview]: https://tug.org/tug2014/slides/mckenna-JSBox.pdf "Overview of JSBox | 2014 Doug McKenna"

----
additional context which may help while answering: 
* my xp in programmatic typesetting is with latex, asciidoc, md/html.
* i like FLOSS.
* my problem with latex is that:
    * it's not very ide-toolable - there're next to none diagnostic tools available for it. the errors are ambiguous in wordings, and hard to resolve
    * it always shows bad bad box overflow error on my template file & i couldn't diagnose what's causing it - nor is it percievable in output pdf.
    * despite using it mostly successfully for well over 18 months, i still am not confident about it... it's based on string replacement and it all feels like a hack. _Just Like_ C preprocessor.
* my use area so far is technical things in maths/computers/electronics: so, equations, flowcharts, code [listing]s, [circuit][circuitikz] and drawings
* and ofcourse the essential things like tables, contents, figures/images, footnotes/references

[listing]: https://www.ctan.org/topic/listing "listing topic on ctan"
[circuitikz]: https://github.com/circuitikz/circuitikz "circuitikz"

----

[t_kk20vlamo]: https://www.youtube.com/watch?v=t_kk20vlamo "Global typesetting with SILE | 2015.07"
[jsbox-trip-test]: https://www.tug.org/TUGboat/tb35-2/tb110mckenna.pdf "Tracing the trip test with JSBox | 2014 Doug McKenna"

**Comment by Omikhleia** [comment]
A1 is easy: no, as far as I know.

I'm afraid the rest would need to be asked to "context-LTMX" or TeX users -- AFAIK, I am not sure I understand any of the underlying questions.

**Comment by Omikhleia** [comment]
Closing the discussion: no feedback since last answer in Nov. 2023.
As noted, it "would need to be asked to "context-LTMX" or TeX users " -- so it is beyond the reach of SILE.
Please feel free to open a new discussion or issue with a better defined scope.

**Comment by alerque** [comment]
This is such a rambling pile of unrelated topics and tangents I don't think there is any way to even dig into it. Even the things brought up for comparison are a hodge-podge of apple slices and orange seeds.

You might be interested in my [Polytype](https://polytype.dev/) project for some direct comparisons between typesetting systems in as close to an apples-to-apples fashion as can be mustered.

If you have specific questions or observations please open one discussion at a time for each distinct topic.
